web-testbed/src/testbed/web_test_harness.py

```diff
@@ -0,0 +1,263 @@
+        if t == "list":
+            return [self._deserialise(i) for i in env["items"]]
+        if t == "tuple":
+            return tuple(self._deserialise(i) for i in env["items"])
+        if t == "dict":
+            out = {}
+            for k_env, v_env in env["items"]:
+                k = self._deserialise(k_env)
+                v = self._deserialise(v_env)
+                out[k] = v
+            return out
+        if t == "time":
+            s = env["value"]
+            h, m, *rest = s.split(":")
+            sec = int(rest[0]) if rest else 0
+            return _dt.time(int(h), int(m), sec)
+        if t == "date":
+            s = env["value"]
+            try:
+                return _dt.datetime.strptime(s, "%m/%d/%Y").date()
+            except ValueError:
+                return _dt.date.fromisoformat(s)
+        # reconstruct functions from source
+        if t == "callable_source":
+            try:
+                scope = {}
+                exec(textwrap.dedent(env["source"]), scope, scope)
+                fn = scope.get(env["name"])
+            except Exception as e:
+                raise ValueError(
+                    f"Failed to exec callable source for {env.get('name')!r}"
+                ) from e
+
+            if not callable(fn):
+                raise ValueError(
+                    f"Callable {env.get('name')!r} not found or not callable after exec"
+                )
+            return fn
+        if t in ("ref", "object"):
+            ref = env.get("ref")
+            if ref is None:
+                ref = env.get("id")
+            return self.my_objs[str(ref)]
+        return env
+
+    def cmd_test_rpc(self, msg):
+        m = msg.to_py() if hasattr(msg, "to_py") else msg
+
+        op = m["op"]
+
+        if op == "getattr":
+            obj = self.my_objs[str(m["obj"])]
+            value = getattr(obj, m["name"])
+            return to_js(
+                self._serialise_payload(value), dict_converter=js.Object.fromEntries
+            )
+
+        if op == "setattr":
+            obj = self.my_objs[str(m["obj"])]
+            setattr(obj, m["name"], self._deserialise(m["value"]))
+            return to_js(
+                self._serialise_payload(None), dict_converter=js.Object.fromEntries
+            )
+
+        if op == "delattr":
+            obj = self.my_objs[str(m["obj"])]
+            delattr(obj, m["name"])
+            return to_js(
+                self._serialise_payload(None), dict_converter=js.Object.fromEntries
+            )
+
+        if op == "call":
+            fn = self.my_objs[str(m["fn"])]
+            args = [self._deserialise(a) for a in m.get("args", [])]
+            kwargs = {k: self._deserialise(v) for k, v in m.get("kwargs", {}).items()}
+            out = fn(*args, **kwargs)
+            return to_js(
+                self._serialise_payload(out), dict_converter=js.Object.fromEntries
+            )
+
+        if op == "hostcall":

```

Comment:
Not sure I understand the difference between "call" and "hostcall"... is that differentiating "call on an object"  and "call a top level/builtin function?"